### PR TITLE
docs: Add sample instructions on multitenancy

### LINKED ISSUE #126: [1] Test and document workload multi-tenancy
To accelerate workload multi-tenancy installation, without creating technical debt, we suggested the following solution:

* "ck8s init" multiple pairs of SC-WC clusters, let's call them P_0, P_1, P_2, ... .
* We configure the same baseDomain and opsDomain for all P_i.
* For P0 we only do "apply sc".
* We point the DNS entry for *.$baseDomain and *.$opsDomain (I still don't get the difference between the two) to one of P_0's SC worker nodes. (I operate under the assumption that XYZ does not have a decent load-balancer.)
* For all P_i with i > 0, we only do "apply wc".

Acceptance criteria:
* Please test if this works and create a Markdown document that shows how to do it. Someone familiar with Compliant Kubernetes should be able to easily follow the prose and copy-paste the code.

### PR DESCRIPTION
**Pull request by OlleLarsson**
**What this PR does / why we need it**:
To get some guidline on how to achive one serivce cluster with multiple workload clusters.

**Which issue this PR fixes**:
fixes #126 

**Special notes for reviewer**:
Don't really know what expected so i guess this is a first draft :)

**Tests**

- Metrics from both workload clusters show up in prometheus/influxDB/grafana
grafana-mc

- Logs from both workload clusters show up in elasticsearch/kibana
logs-cluster-label

**Checklist:**

- [ ] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [ ] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

### CODE REVIEW

**Review by Xartos — COMMENTED** [review]
I think this is great! Maybe mention something like
```
This documentation is assuming you are setting up a cluster on Exoscale.
Although the same principles apply on the other clouds
```

And another thing.
Should this file live in root? or should we have some docs folder we can put this in?

**Comment by OlleLarsson** [comment]
> I think this is great! Maybe mention something like
> 
> ```
> This documentation is assuming you are setting up a cluster on Exoscale.
> Although the same principles apply on the other clouds
> ```
> 
> And another thing.
> Should this file live in root? or should we have some docs folder we can put this in?

We can perhpas simply put it in `docs/`?

File: multitenancy.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+### Infrastructure and kubernetes
+1. Set some variables
+  ```bash
+  domain=<your_domain>
+  pgp_fp=<your_pgp_fp>
+  config_path=<path_to_store_config_repos>
+  ```
+
+1. Init repos
+  ```bash
```
Comment (cristiklein) [inline multitenancy.md:25]: Can you tell how `ck8s_linux_amd64` can be found?

File: multitenancy.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+# Remove dns records
+exo dns remove a1ck.io *.ops.mtt-0
+exo dns remove a1ck.io *.mtt-0
+```
+
+## Unknowns and failuers
+
+- OIDC and kubelogin has not been tested.
+
+- The certificate for the kubeapi-metrics ingress will never be ready in the workload clusters as all clusters share the same `baseDomain` and `opsDomain` which are both pointing to the service cluster.
```
Comment (cristiklein) [inline multitenancy.md:92]: I don't understand this. :( Refresh my memory, what is the `kubeapi-metrics` Ingress and why is it neded? What are the consequences of this not working?

File: multitenancy.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+- Have not used ingress to any of the workload clusters.
```
Comment (cristiklein) [inline multitenancy.md:94]: Can you use the same tense as above? "Ingress of workload clusters has not been tested, but is expected to work without issues."

**Review by cristiklein — APPROVED** [review]
This is awesome!

LGTM once the minor comments are addressed.

I would very much love to have this here, to get started with operator documentation: https://compliantkubernetes.io/operator-manual/lifecycle/

File: multitenancy.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# Multitenancy
+
+This document contains sample instructions on how to setup one service cluster together with multiple workload clusters. As stated, the instructions below are just samples, you need update them according to your situation and requirements. You will find more details on the commands executed in the cluster and apps repos.
+
+The following repositores and commits have been tested
+- [ck8s-cluster](https://github.com/elastisys/ck8s-cluster/tree/324ac074a337b99b83a464ce0189ed2c0f267e19)
+- [compliantkubernetes-apps](https://github.com/elastisys/compliantkubernetes-apps)
```
Comment (viktor-f) [inline multitenancy.md:7]: The link to the apps repo is pointing to `main`. In the text above it seems like you say that this should point to the specific commit that you used.

File: multitenancy.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  export CK8S_ENVIRONMENT_NAME=mtt-2
+  export CK8S_CONFIG_PATH=${config_path}/mtt-2
+  ./bin/ck8s init
+  ```
+
+2. Edit
+  - `secrets.yaml` for all clusters
+  - `sc-config.yaml` for the service cluster
+  - `wc-config.yaml` for all workload cluster
+  You should set the same `opsDomain` and `baseDomain` in all clusters.
```
Comment (viktor-f) [inline multitenancy.md:73]: Did you try changing the `baseDomain` of the workload cluster to something else? If that works then we could have a wildcard DNS record pointing to each workload cluster as well.

File: multitenancy.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+### Teardown
```
Comment (viktor-f) [inline multitenancy.md:75]: You are missing the step where you apply apps

File: multitenancy.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+3. Get ip of sc loadbalancer
+  ```bash
+  sc_lb_ip=$(./dist/ck8s_linux_amd64 internal terraform output --cluster sc --config-path ${config_path}/mtt-0 | jq -r '.sc_ingress_controller_lb_ip_address.value')
+  ```
+
+4. Create DNS records
+  ```bash
+  exo dns add A ${domain} -a ${sc_lb_ip} -n *.ops.mtt-0
+  exo dns add A ${domain} -a ${sc_lb_ip} -n *.mtt-0
```
Comment (viktor-f) [inline multitenancy.md:48]: I don't think we have the `exo` cli as a requirement anywhere. So either add a link to where one can download it or instead use our `ck8s-dns` repo.

File: multitenancy.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+```bash
+# Tear down infra and k8s
+./dist/ck8s_linux_amd64 destroy --cluster sc --config-path ${config_path}/mtt-0 --kubernetes-cleanup=false
+./dist/ck8s_linux_amd64 destroy --cluster wc --config-path ${config_path}/mtt-1 --kubernetes-cleanup=false
+./dist/ck8s_linux_amd64 destroy --cluster wc --config-path ${config_path}/mtt-2 --kubernetes-cleanup=false
+
+# Remove dns records
+exo dns remove a1ck.io *.ops.mtt-0
+exo dns remove a1ck.io *.mtt-0
```
Comment (viktor-f) [inline multitenancy.md:85]: Same thing here regarding `exo`.

**Review by viktor-f — CHANGES_REQUESTED** [review]
Looks nice! Found some comments

File: multitenancy.md (hunk 6 above)
Comment (OlleLarsson) [reply multitenancy.md:75]: Correct

File: multitenancy.md (hunk 7 above)
Comment (OlleLarsson) [reply multitenancy.md:48]: ck8s-dns is not an open project. I'll add a link to exo cli

File: multitenancy.md (hunk 1 above)
Comment (OlleLarsson) [reply multitenancy.md:25]: In ck8s-cluster when you have built the cli
```
cd ck8s-cluster && make
ls dist/
```

File: multitenancy.md (hunk 4 above)
Comment (OlleLarsson) [reply multitenancy.md:7]: Yes was waiting for this to be approved and merged https://github.com/elastisys/compliantkubernetes-apps/pull/145

File: multitenancy.md (hunk 5 above)
Comment (OlleLarsson) [reply multitenancy.md:73]: I'll try and check :)

File: multitenancy.md (hunk 7 above)
Comment (viktor-f) [reply multitenancy.md]: Good point, forgot about that.

File: multitenancy.md (hunk 2 above)
Comment (OlleLarsson) [reply multitenancy.md:113]: Got the certificates to work by pointing dns to the clusters.
The kubeapi-metrics is used to expose the `/healtz` endpoint of the kubeapi-server. This endpoint is normally checked by the blackbox exporter in the service cluster.

**Comment by OlleLarsson** [comment]
I think I've addressed all you comments.
@cristiklein should I move this pr to https://github.com/elastisys/compliantkubernetes ?

File: multitenancy.md (hunk 5 above)
Comment (OlleLarsson) [reply multitenancy.md]: Pointed dns to wc loadbalancer and used unique basedomain for each cluster and it seems to work. Now only opsdomain is shared between the clusters.

**Comment by cristiklein** [comment]
@OlleLarsson Yes, please. We really lack documentation in this section https://compliantkubernetes.io/operator-manual/lifecycle/

**Comment by OlleLarsson** [comment]
> @OlleLarsson Yes, please. We really lack documentation in this section https://compliantkubernetes.io/operator-manual/lifecycle/

https://github.com/elastisys/compliantkubernetes/pull/35

**Pull request closed without merging**